README

dvdns: An authoritative-only DNS server.




Specific notes:
* SQL rdata must be in wire format, inside the database


Core dependencies:
* GLib 2.x	http://www.gtk.org/
* GNet 2.x	http://www.gnetlibrary.org/
* SQLite 3.x	http://www.sqlite.org/


Perl dependencies (needed for import, not for normal operation):
* DBI
* DBD::SQLite
* Net::DNS::RR
* Net::DNS::ZoneFile::Fast
* ...and their many dependencies.



NEWS

Release 0.2 -- Not yet released

- ...


Release 0.1 -- November 8, 2006

- Initial release.  Basic queries work.



dnsd.h

/*
 * Copyright 2006 Jeff Garzik
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#ifndef __DNSD_H__
#define __DNSD_H__

#include <stdint.h>
#include <time.h>
#include <glib.h>

#define ARRAY_SIZE(x) (sizeof(x) / sizeof((x)[0]))

enum {
	max_label_len		= 63,
	initial_name_alloc	= 512,

	qtype_all		= 255,

	rcode_nxdomain		= 3,
	rcode_notimpl		= 4,

	op_query		= 0,
};

enum blob_hash_init_info {
	BLOB_HASH_INIT		= 5381UL
};

struct dns_msg_hdr {
	uint16_t		id;
	unsigned char		opts[2];
	uint16_t		n_q;
	uint16_t		n_ans;
	uint16_t		n_auth;
	uint16_t		n_add;
};

enum dns_hdr_bits {
	hdr_response		= 1 << 7,
	hdr_auth		= 1 << 2,
	hdr_req_recur		= 1 << 0,
	hdr_opcode_mask		= 0x78,
	hdr_opcode_shift	= 3,
};

struct dnsq {
	GList			*labels;
	unsigned int		type;
	unsigned int		class;

	char			*name;
	unsigned int		name_alloc;
};

struct dnsres {
	char			*buf;
	unsigned int		buflen;
	unsigned int		alloc_len;
	unsigned int		hdrq_len;
	GList			*queries;
	int			query_rc;

	unsigned int		n_answers;
	unsigned int		n_refs;

	time_t			mc_expire;		/* cache expiration time */
	unsigned long		hash;		/* raw message hash */
};

struct backend_rr {
	const unsigned char	*domain;
	int			type;
	int			class;
	int			ttl;
	const void		*rdata;
	unsigned int		rdata_len;
};

struct dns_server_stats {
	unsigned long		sql_q;		/* SQL queries */
	unsigned long		udp_q;		/* UDP queries */
	unsigned long		tcp_q;		/* TCP queries */
	unsigned long		mc_hit;		/* msg cache hits */
	unsigned long		mc_miss;	/* msg cache misses */
};

/* backend.c */
extern void backend_init(void);
extern void backend_exit(void);
extern void backend_query(const struct dnsq *, struct dnsres *);

/* dns.c */
static inline struct dnsres *dnsres_ref(struct dnsres *res)
{
	res->n_refs++;
	return res;
}
extern void dnsres_unref(struct dnsres *res);
extern struct dnsres *dns_message(const char *buf, unsigned int buflen);
extern void dns_push_rr(struct dnsres *res, const struct backend_rr *rr);
extern void dns_set_rcode(struct dnsres *res, unsigned int code);
extern void dns_init(void);

/* socket.c */
extern void init_net(void);

/* main.c */
extern int dns_port;
extern char db_fn[];
extern struct dns_server_stats srvstat;

#endif /* __DNSD_H__ */


main.c

/*
 * Copyright 2006 Jeff Garzik
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <syslog.h>
#include <errno.h>
#include <string.h>
#include <glib.h>
#include <gnet.h>
#include "dnsd.h"

char db_fn[4096] = "dns.db";
char pid_fn[4096] = "dvdnsd.pid";
int dns_port = 9953;
static int foreground;
struct dns_server_stats srvstat;

static void show_usage(const char *prog)
{
	fprintf(stderr, "usage: %s [options]\n"
		"options:\n"
		"  -f FILE		use sqlite database FILE\n"
		"  -p PORT		bind to port PORT\n"
		"  -P FILE		Write daemon process id to FILE\n",
		prog);
	exit(1);
}

static void parse_cmdline(int argc, char **argv)
{
	int opt;

	while ((opt = getopt(argc, argv, "hf:Fp:P:")) != -1) {
		switch (opt) {
			case 'f':
				strcpy(db_fn, optarg);
				break;
			case 'F':
				foreground = 1;
				break;
			case 'p':
				if (atoi(optarg) > 0 &&
				    atoi(optarg) < 65536)
					dns_port = atoi(optarg);
				else {
					fprintf(stderr, "invalid DNS port %s\n",
						optarg);
					exit(1);
				}
				break;
			case 'P':
				strcpy(pid_fn, optarg);
				break;
			default:
				show_usage(argv[0]);
				exit(1);
		}
	}
}

static void syslogerr(const char *prefix)
{
	syslog(LOG_ERR, "%s: %s", prefix, strerror(errno));
}

static void write_pid_file(void)
{
	char str[32], *s;
	size_t bytes;

	sprintf(str, "%u\n", getpid());
	s = str;
	bytes = strlen(s);

	int fd = open(pid_fn, O_WRONLY | O_CREAT | O_EXCL, 0666);
	if (fd < 0) {
		syslogerr("open pid");
		exit(1);
	}

	while (bytes > 0) {
		ssize_t rc = write(fd, s, bytes);
		if (rc < 0) {
			syslogerr("write pid");
			exit(1);
		}

		bytes -= rc;
		s += rc;
	}

	if (close(fd) < 0)
		syslogerr("close pid");
}

int main (int argc, char *argv[])
{
	GMainLoop *loop;

	memset(&srvstat, 0, sizeof(srvstat));

	parse_cmdline(argc, argv);

	openlog("dvdnsd", LOG_PID, LOG_LOCAL3);

	if ((!foreground) && (daemon(1, 0) < 0)) {
		syslogerr("daemon");
		return 1;
	}

	write_pid_file();

	gnet_init();

	loop = g_main_loop_new(NULL, FALSE);
	g_assert(loop != NULL);

	init_net();
	backend_init();
	dns_init();

	syslog(LOG_INFO, "initialized");

	g_main_loop_run(loop);

	backend_exit();

	return 0;
}



socket.c

/*
 * Copyright 2006 Jeff Garzik
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <glib.h>
#include <gnet.h>
#include "dnsd.h"

enum client_state {
	idle,
	msglen,
	msg,
};

struct client {
	GConn			 *conn;
	enum client_state	state;
};

static GUdpSocket *udpsock;


static void udp_message(GUdpSocket *sock, const GInetAddr *src,
			const char *buf, unsigned int buflen)
{
	struct dnsres *res = dns_message(buf, buflen);

	srvstat.udp_q++;

	if (res) {
		gnet_udp_socket_send(sock, res->buf, res->buflen, src);
		dnsres_unref(res);
	}
}

static gboolean udp_rx (GIOChannel *source, GIOCondition condition,
                                             void *data)
{
	char buf[2048];
	int bytes;
	GInetAddr *src = NULL;

	bytes = gnet_udp_socket_receive(udpsock, buf, sizeof(buf), &src);
	g_assert (bytes > 0);
	g_assert (src != NULL);

	udp_message(udpsock, src, buf, bytes);

	return TRUE; /* poll again */
}

static void tcp_message(struct client *cli, const char *buf, unsigned int buflen)
{
	struct dnsres *res = dns_message(buf, buflen);

	srvstat.tcp_q++;

	if (res) {
		uint16_t msglen = g_htons(res->buflen);
		gnet_conn_write(cli->conn, (gchar *) &msglen, 2);
		gnet_conn_write(cli->conn, res->buf, res->buflen);
		dnsres_unref(res);
	}
}

static void cli_close(struct client *cli)
{
	gnet_conn_unref(cli->conn);
	g_slice_free(struct client, cli);
}

static void tcp_conn (GConn *conn, GConnEvent *event, void *user_data)
{
	struct client *cli = user_data;

	switch (event->type) {
	case GNET_CONN_ERROR:
	case GNET_CONN_CLOSE:
		cli_close(cli);
		break;

	case GNET_CONN_WRITE:
		break;

	case GNET_CONN_READ:
		if (cli->state == msglen) {
			unsigned short *s = (unsigned short *) event->buffer;
			int msglen = g_ntohs(*s);

			gnet_conn_readn(cli->conn, msglen);
			cli->state = msg;
		}
		else if (cli->state == msg) {
			tcp_message(cli, event->buffer, event->length);

			gnet_conn_readn(cli->conn, 2);
			cli->state = msglen;
		}
		break;

	default:
		/* do nothing */
		break;
	}
}

static void tcp_accept (GServer *server, GConn *client, void *data)
{
	struct client *cli;

	g_assert(client != NULL);	/* socket error */

	cli = g_slice_new0(struct client);
	g_assert(cli != NULL);

	cli->conn = client;
	cli->state = idle;

	gnet_conn_set_callback(client, tcp_conn, cli);

	gnet_conn_readn(client, 2);
	cli->state = msglen;
}

void init_net(void)
{
	GIOChannel *udpchan;
	GServer *tcpsrv;

	udpsock = gnet_udp_socket_new_with_port (dns_port);
	g_assert(udpsock != NULL);

	udpchan = gnet_udp_socket_get_io_channel (udpsock);
	g_assert(udpchan != NULL);

	g_io_add_watch(udpchan, G_IO_IN, udp_rx, NULL);

	tcpsrv = gnet_server_new(NULL, dns_port, tcp_accept, NULL);
	g_assert(tcpsrv != NULL);
}



backend.c

/*
 * Copyright 2006 Jeff Garzik
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <sqlite3.h>
#include "dnsd.h"

enum sql_stmt_indices {
	st_name,

	st_last = st_name
};

static const char *sql_stmt_text[] = {
	/* st_name */
	"select labels.name, rrs.* from labels, rrs where "
	"labels.id = rrs.domain and "
	"rrs.domain in "
	"(select labels.id from labels where labels.name = ?)",
};

static sqlite3_stmt *prep_stmts[st_last + 1];
static sqlite3 *db;

void backend_init(void)
{
	unsigned int i;
	int rc;

	rc = sqlite3_open(db_fn, &db);
	if (rc != SQLITE_OK) {
		syslog(LOG_ERR, "sqlite3_open failed");
		exit(1);
	}

	for (i = 0; i <= st_last; i++) {
		const char *dummy;

		rc = sqlite3_prepare(db, sql_stmt_text[i],
				     strlen(sql_stmt_text[i]),
				     &prep_stmts[i], &dummy);
		g_assert(rc == SQLITE_OK);
	}
}

void backend_exit(void)
{
	unsigned int i;
	int rc;

	for (i = 0; i <= st_last; i++)
		sqlite3_finalize(prep_stmts[i]);

	rc = sqlite3_close(db);
	g_assert(rc == SQLITE_OK);
}

void backend_query(const struct dnsq *q, struct dnsres *res)
{
	int rc;
	unsigned int idx, rows = 0;

	idx = st_name;

	rc = sqlite3_bind_text(prep_stmts[idx], 1,
			      q->name, strlen(q->name),
			      SQLITE_STATIC);
	g_assert(rc == SQLITE_OK);

	while (1) {
		struct backend_rr rr;

		srvstat.sql_q++;

		/* execute SQL query */
		rc = sqlite3_step(prep_stmts[idx]);
		if (rc == SQLITE_DONE || rc == SQLITE_BUSY)
			break;
		g_assert(rc == SQLITE_ROW);

		rows++;

		memset(&rr, 0, sizeof(rr));
		rr.domain = sqlite3_column_text(prep_stmts[idx], 0);
		/* skip suffix, column #1 */
		rr.type = sqlite3_column_int(prep_stmts[idx], 2);
		rr.class = sqlite3_column_int(prep_stmts[idx], 3);
		rr.ttl = sqlite3_column_int(prep_stmts[idx], 4);
		rr.rdata = sqlite3_column_blob(prep_stmts[idx], 5);
		rr.rdata_len = sqlite3_column_bytes(prep_stmts[idx], 5);

		/* filter out non-matching classes and types */
		if (q->class != rr.class)
			continue;
		if ((q->type != qtype_all) && (q->type != rr.type))
			continue;

		dns_push_rr(res, &rr);
	}

	rc = sqlite3_reset(prep_stmts[idx]);
	g_assert(rc == SQLITE_OK);

	/* no data found for given domain name */
	if (rows == 0)
		dns_set_rcode(res, rcode_nxdomain);
}



import-zone.pl
#!/usr/bin/perl -w
#
# Script to import a standard BIND zone file into an SQL database.
#
#
# Copyright 2006 Jeff Garzik
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; see the file COPYING.  If not, write to
# the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
#

use Net::DNS::RR;
use Net::DNS::ZoneFile::Fast;
use DBI qw(:sql_types);

my ($dbh, %dom_cache, $next_id);

sub usage() {
	print STDERR "usage: import-zone.pl DATABASE ZONE-FILE\n";
	exit(1);
}

sub read_max_id() {
	my $sth = $dbh->prepare("select id from labels " .
				"order by id desc " .
				"limit 1");
	die "select id prep died\n" unless $sth;

	$sth->execute() or die "select id exec died";

	my (@data, $max_id);
	if (@data = $sth->fetchrow_array()) {
		$max_id = $data[0];
		die "invalid max_id $max_id\n"
			unless ($max_id > 0);

		$next_id = $max_id + 1;
	} else {
		$next_id = 1;
	}
}

sub get_dom_id($) {
	my ($domain) = @_;

	return $dom_cache{$domain}
		if (exists $dom_cache{$domain});

	# obtain integer id associated with domain name, or create new one
	my $sth = $dbh->prepare('select id from labels where name = ?');
	die "select prep failed" unless $sth;

	$sth->execute($domain) or die "sql exec failed";

	if (($id) = $sth->fetchrow_array()) {
		$dom_cache{$domain} = $id;
		return $id;
	}

	return undef;
}

sub import_rr($) {
	my ($rr) = @_;
	my ($id);

	# get domain integer id, or create new one
	$id = get_dom_id($rr->name);
	if (!$id) {
		# create new id
		$id = $next_id;
		$next_id++;

		# store in cache
		$dom_cache{$rr->name} = $id;

		# store in database
		my $ih = $dbh->prepare('insert into labels values (?,?)');
		$ih->execute($rr->name, $id) or die "sql insert failed";
	}

	# build RR sql insert
	my $sth = $dbh->prepare_cached('insert into rrs values (?,?,?,?,?)');
	die "sql prep failed" unless $sth;

	$sth->bind_param(1, $id, SQL_INTEGER);
	$sth->bind_param(2, Net::DNS::typesbyname($rr->type), SQL_INTEGER);
	$sth->bind_param(3, Net::DNS::classesbyname($rr->class), SQL_INTEGER);
	$sth->bind_param(4, $rr->ttl, SQL_INTEGER);
	$sth->bind_param(5, $rr->_canonicalRdata, SQL_BLOB);

	$sth->execute() or die "sql exec failed";
}

sub import_zonefile($) {
	my ($fn) = @_;
	my ($rrs, $rr);

	$rrs = Net::DNS::ZoneFile::Fast::parse(
		'file'		=> $fn,
		'tolower'	=> 1
		);

	foreach $rr (@$rrs) {
		import_rr($rr);
	}

	$dbh->commit;
}


my $dbfn = shift;
usage() unless $dbfn;
$dbh = DBI->connect("dbi:SQLite:dbname=$dbfn", "", "",
		    { AutoCommit => 0 });
$dbh->{unicode} = 1;
die "connect($dbfn) failed: " . DBI->errstr . "\n"
	unless $dbh;

read_max_id();
$dom_cache{""} = 0;

my ($zonefn);
while ($zonefn = shift) {
	import_zonefile($zonefn);
}

$dbh->disconnect;

exit(0);



dns.c

/*
 * Copyright 2006 Jeff Garzik
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <string.h>
#include <time.h>
#include <glib.h>
#include "dnsd.h"

enum {
	MSG_CACHE_EXPIRE		= 60,
};

static GQueue		*msg_expire_q;
static GHashTable	*msg_cache;
static time_t		current_time;


/* "djb2"-derived hash function */
static unsigned long blob_hash(unsigned long hash, const void *_buf, size_t buflen)
{
	const unsigned char *buf = _buf;
	int c;

	while (buflen > 0) {
		c = *buf++;
		buflen--;

		hash = ((hash << 5) + hash) ^ c; /* hash * 33 ^ c */
	}

	return hash;
}

static void msg_cache_expire(void)
{
	struct dnsres *res;

	while (1) {
		res = g_queue_peek_head(msg_expire_q);
		if (!res)
			return;
		if (current_time < res->mc_expire)
			return;

		g_queue_pop_head(msg_expire_q);
		g_hash_table_remove(msg_cache, (gpointer) res->hash);
	}
}

static struct dnsres *msg_cache_lookup(const char *buf, unsigned int buflen,
				       int *expired, unsigned long *hash_out)
{
	struct dnsres *res;
	unsigned long hash;

	*expired = 0;

	hash = blob_hash(BLOB_HASH_INIT, buf, buflen);
	hash = blob_hash(hash, &buflen, sizeof(buflen));
	*hash_out = hash;

	res = g_hash_table_lookup(msg_cache, (gpointer) hash);
	if (!res)
		return NULL;

	if (current_time < res->mc_expire)
		return res;
	
	msg_cache_expire();
	*expired = 1;
	return NULL;
}

static void msg_cache_add(unsigned long hash, struct dnsres *res)
{
	g_hash_table_insert(msg_cache, (gpointer) hash, res);
	g_queue_push_tail(msg_expire_q, res);

}

void dns_set_rcode(struct dnsres *res, unsigned int code)
{
	struct dns_msg_hdr *hdr;

	hdr = (struct dns_msg_hdr *) res->buf;
	hdr->opts[1] = code & 0x0f;
}

static void dns_finalize(struct dnsres *res)
{
	struct dns_msg_hdr *hdr;

	hdr = (struct dns_msg_hdr *) res->buf;
	hdr->n_ans = g_htons(res->n_answers);
	hdr->opts[0] |= hdr_auth;
}

static void dns_res_grow(struct dnsres *res, unsigned int buflen)
{
	size_t new_size = res->alloc_len;
	void *mem;

	do {
		new_size = new_size << 1;
	} while ((new_size - res->buflen) < buflen);

	mem = g_slice_alloc(new_size);
	memcpy(mem, res->buf, res->alloc_len);
	g_slice_free1(res->alloc_len, res->buf);

	res->buf = mem;
	res->alloc_len = new_size;
}

static void dns_push_bytes(struct dnsres *res, const void *buf,
			   unsigned int buflen)
{
	if ((res->alloc_len - res->buflen) < buflen)
		dns_res_grow(res, buflen);

	memcpy(res->buf + res->buflen, buf, buflen);
	res->buflen += buflen;
}

void dns_push_rr(struct dnsres *res, const struct backend_rr *rr)
{
	const unsigned char *s, *accum;
	size_t accum_len;
	uint32_t ttl;
	uint16_t tmp;
	uint8_t zero8 = 0;

	accum_len = 0;
	accum = s = rr->domain;
	while (1) {
		if (*s == '.' || *s == 0) {
			if (accum_len) {
				uint8_t len = accum_len;
				g_assert(len <= 63);
				dns_push_bytes(res, &len, 1);
				dns_push_bytes(res, accum, accum_len);
			}

			if (*s == 0)
				break;

			accum = s + 1;
			accum_len = 0;
		} else
			accum_len++;

		s++;
	}

	dns_push_bytes(res, &zero8, 1);

	tmp = g_htons(rr->type);
	dns_push_bytes(res, &tmp, 2);

	tmp = g_htons(rr->class);
	dns_push_bytes(res, &tmp, 2);

	ttl = g_htonl(rr->ttl);
	dns_push_bytes(res, &ttl, 4);

	tmp = g_htons(rr->rdata_len);
	dns_push_bytes(res, &tmp, 2);
	dns_push_bytes(res, rr->rdata, rr->rdata_len);

	res->n_answers++;
}

static void list_free_ent(void *data, void *user_data)
{
	g_free(data);
}

static void dnsres_free_q(void *data, void *user_data)
{
	struct dnsq *q = data;

	g_list_foreach(q->labels, list_free_ent, NULL);
	g_list_free(q->labels);
	g_slice_free1(q->name_alloc, q->name);
	g_slice_free(struct dnsq, q);
}

static void dnsres_free(struct dnsres *res)
{
	g_list_foreach(res->queries, dnsres_free_q, NULL);
	g_list_free(res->queries);
	g_slice_free1(res->alloc_len, res->buf);
	g_slice_free(struct dnsres, res);
}

void dnsres_unref(struct dnsres *res)
{
	g_assert(res->n_refs > 0);

	res->n_refs--;
	if (!res->n_refs)
		dnsres_free(res);
}

static struct dnsres *dnsres_alloc(void)
{
	struct dnsres *res = g_slice_new0(struct dnsres);
	if (res)
		res->n_refs = 1;
	return res;
}

static void dnsq_append_label(struct dnsq *q, const char *buf, unsigned int buflen)
{
	char *label;

	/* create label record */
	label = g_utf8_strdown(buf, buflen);

	/* add label to question's list of labels */
	q->labels = g_list_append(q->labels, label);

	if (!q->name) {
		q->name = g_slice_alloc(initial_name_alloc);
		q->name[0] = 0;
		q->name_alloc = initial_name_alloc;
	}
	else if ((buflen+1) > (q->name_alloc - strlen(q->name))) {
		void *mem;
		size_t new_size = q->name_alloc << 1;

		/* simulate g_realloc with slices */
		mem = g_slice_alloc(new_size);
		memcpy(mem, q->name, q->name_alloc);
		g_slice_free1(q->name_alloc, q->name);

		q->name = mem;
		q->name_alloc = new_size;
	}

	if (q->name[0] != 0)
		strcat(q->name, ".");
	strcat(q->name, label);
}

enum {
	max_ptr_stack		= 32,
};

struct ptr_stack {
	unsigned int		len;
	unsigned int		ptr[max_ptr_stack];
};

static int in_ptr_history(struct ptr_stack *ps, unsigned int ptr)
{
	unsigned int len = ps->len;
	unsigned int i;

	for (i = 0; i < len; i++)
		if (ps->ptr[i] == ptr)
			return 1;

	if (len == max_ptr_stack)
		return 1;

	ps->ptr[len] = ptr;
	ps->len++;

	return 0;
}

static int dns_parse_label(struct dnsq *q, const char *label,
			   const char *msg, unsigned int msg_len)
{
	unsigned int idx = label - msg;
	unsigned int free, saved_len = 0, ptr_chasing = 0;
	const char *p;
	struct ptr_stack ps;

	ps.len = 0;

	/* read list of labels */
next_ptr:
	p = &msg[idx];
	free = msg_len - idx;
	while (1) {
		unsigned int label_len, label_flags;

		if (free == 0)
			goto err_out;

		/* get label length */
		label_len = *p;
		p++;
		free--;

		/* move bits 7-6 to label_flags */
		label_flags = label_len & 0xc0;
		label_len &= 0x3f;

		/* bits 01 and 10 are reserved / not handled */
		if (label_flags == 0x80 || label_flags == 0x40)
			goto err_out;

		/* pointer compression (offset-based labels) */
		if (label_flags == 0xc0) {
			if (free == 0)
				goto err_out;

			idx = (label_len << 8) | (*p);
			p++;
			free--;

			if (idx >= msg_len)
				goto err_out;
			if (in_ptr_history(&ps, idx))
				goto err_out;

			if (!ptr_chasing) {
				ptr_chasing = 1;
				saved_len = p - label;
			}

			goto next_ptr;
		}

		/* normal label */
		else {
			/* verify free space */
			if (label_len > free)
				goto err_out;

			/* if label length zero, list terminates */
			if (label_len == 0)
				break;

			/* copy label */
			dnsq_append_label(q, p, label_len);

			p += label_len;
			free -= label_len;
		}
	}

	if (ptr_chasing)
		return saved_len;
	return p - label;

err_out:
	return -1;
}

static int dns_parse_msg(struct dnsres *res, const struct dns_msg_hdr *hdr,
			 const char *msg, unsigned int msg_len)
{
	unsigned int i;
	const char *ibuf = msg;
	unsigned int ibuflen = msg_len;
	unsigned int n_q = g_ntohs(hdr->n_q);
	int rc = 0;

	ibuf += sizeof(*hdr);
	ibuflen -= sizeof(*hdr);

	for (i = 0; i < n_q; i++) {
		struct dnsq *q;
		uint16_t tmpi;
		int label_len;

		q = g_slice_new0(struct dnsq);
		g_assert(q != NULL);

		/*
		 * read label, with pointer decompression
		 */
		label_len = dns_parse_label(q, ibuf, msg, msg_len);
		if (label_len < 0)
			goto err_out;

		ibuf += label_len;
		ibuflen -= label_len;

		/*
		 * read type, class
		 */
		if (ibuflen < 4)
			goto err_out;

		memcpy(&tmpi, ibuf, 2);
		q->type = g_ntohs(tmpi);

		memcpy(&tmpi, ibuf + 2, 2);
		q->class = g_ntohs(tmpi);

		ibuf += 4;
		ibuflen -= 4;

		/* add to list of queries */
		res->queries = g_list_append(res->queries, q);
	}

out:
	/* note length of hdr + query section */
	res->hdrq_len = msg_len - ibuflen;
	return rc;

err_out:
	rc = -1;
	goto out;
}

struct dnsres *dns_message(const char *buf, unsigned int buflen)
{
	const struct dns_msg_hdr *hdr;
	struct dns_msg_hdr *ohdr;
	struct dnsres *res;
	char *obuf;
	unsigned int opcode;
	unsigned long hash;
	int rc, expired;

	current_time = time(NULL);

	/* look up raw bytes in message cache */
	res = msg_cache_lookup(buf, buflen, &expired, &hash);
	if (res) {
		srvstat.mc_hit++;
		return dnsres_ref(res);
	}

	srvstat.mc_miss++;

	/* allocate result struct */
	res = dnsres_alloc();
	if (!res)
		return NULL;

	res->mc_expire = current_time + MSG_CACHE_EXPIRE;

	/* bail, if packet smaller than dns header */
	if (buflen < sizeof(*hdr))
		goto err_out;

	hdr = (const struct dns_msg_hdr *) buf;

	/* if this is a response packet, just return NULL (no resp to client) */
	if (hdr->opts[0] & hdr_response)
		goto err_out;

	/* read list of questions */
	rc = dns_parse_msg(res, hdr, buf, buflen);
	if (rc != 0)			/* invalid input */
		goto err_out;

	/* allocate output buffer */
	res->alloc_len = MAX(1024, buflen);
	obuf = res->buf = g_slice_alloc(res->alloc_len);
	g_assert(obuf != NULL);

	/* copy hdr + query section into response packet */
	memcpy(obuf, buf, res->hdrq_len);
	res->buflen = res->hdrq_len;

	/* sanitize response header */
	ohdr = (struct dns_msg_hdr *) obuf;
	ohdr->opts[0] =
		hdr_response |
		(hdr->opts[0] & hdr_opcode_mask) |
		(hdr->opts[0] & hdr_req_recur);
	ohdr->n_ans = 0;
	ohdr->n_auth = 0;
	ohdr->n_add = 0;

	opcode = (hdr->opts[0] & hdr_opcode_mask) >> hdr_opcode_shift;
	switch (opcode) {
		case op_query:
			g_list_foreach(res->queries,
				       (GFunc) backend_query, res);
			if (res->query_rc != 0)		/* query failed */
				goto err_out;
			break;

		default:
			dns_set_rcode(res, rcode_notimpl);
			break;
	}

	dns_finalize(res);

	/* add to message cache */
	if (!expired)
		msg_cache_expire();
	msg_cache_add(hash, dnsres_ref(res));

	return res;

err_out:
	dnsres_unref(res);
	return NULL;
}

void dns_init(void)
{
	msg_cache = g_hash_table_new_full(g_direct_hash, g_direct_equal,
					  NULL, (GDestroyNotify) dnsres_unref);
	g_assert(msg_cache != NULL);

	msg_expire_q = g_queue_new();
	g_assert(msg_expire_q != NULL);
}

